nifi/nifi-nar-bundles/nifi-kite-bundle/nifi-kite-processors/src/main/java/org/apache/nifi/processors/kite/ConvertAvroSchema.java

```diff
@@ -0,0 +1,252 @@
+			.<Relationship> builder().add(SUCCESS).add(FAILURE)
+			.add(INCOMPATIBLE).build();
+
+	@Override
+	protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+		return PROPERTIES;
+	}
+
+	@Override
+	public Set<Relationship> getRelationships() {
+		return RELATIONSHIPS;
+	}
+
+	@Override
+	public void onTrigger(ProcessContext context, final ProcessSession session)
+			throws ProcessException {
+		FlowFile incomingAvro = session.get();
+		if (incomingAvro == null) {
+			return;
+		}
+
+		String inputSchemaProperty = context.getProperty(INPUT_SCHEMA)
+				.evaluateAttributeExpressions(incomingAvro).getValue();
+		final Schema inputSchema;
+		try {
+			inputSchema = getSchema(inputSchemaProperty,
+					DefaultConfiguration.get());
+		} catch (SchemaNotFoundException e) {
+			getLogger().error("Cannot find schema: " + inputSchemaProperty);
+			session.transfer(incomingAvro, FAILURE);
+			return;
+		}
+		String outputSchemaProperty = context.getProperty(OUTPUT_SCHEMA)
+				.evaluateAttributeExpressions(incomingAvro).getValue();
+		final Schema outputSchema;
+		try {
+			outputSchema = getSchema(outputSchemaProperty,
+					DefaultConfiguration.get());
+		} catch (SchemaNotFoundException e) {
+			getLogger().error("Cannot find schema: " + outputSchemaProperty);
+			session.transfer(incomingAvro, FAILURE);
+			return;
+		}
+		String fieldMappingProperty = context.getProperty(FIELD_MAPPING)
+				.evaluateAttributeExpressions(incomingAvro).getValue();
+		final JsonNode fieldMapping;
+		try {
+			fieldMapping = getJSON(fieldMappingProperty,
+					DefaultConfiguration.get());
+		} catch (IllegalArgumentException e) {
+			getLogger().error(
+					"Cannot find field mapping: " + fieldMappingProperty);
+			session.transfer(incomingAvro, FAILURE);
+			return;
+		}
+		final AvroRecordConverter converter = new AvroRecordConverter(
+				inputSchema, outputSchema, fieldMapping);
+
+		final DataFileWriter<Record> writer = new DataFileWriter<>(
+				AvroUtil.newDatumWriter(outputSchema, Record.class));
+		writer.setCodec(CodecFactory.snappyCodec());
+
+		try {
+			final LongHolder written = new LongHolder(0L);
+			final FailureTracker failures = new FailureTracker();
+
+			FlowFile badRecords = session.clone(incomingAvro);
+			FlowFile outgoingAvro = session.write(incomingAvro,
+					new StreamCallback() {
+						@Override
+						public void process(InputStream in, OutputStream out)
+								throws IOException {
+							try (DataFileStream<Record> stream = new DataFileStream<Record>(
+									in, new GenericDatumReader<Record>(
+											converter.getInputSchema()))) {
+								try (DataFileWriter<Record> w = writer.create(
+										outputSchema, out)) {
+									for (Record record : stream) {
+										try {
+											Record converted = converter
+													.convert(record);
+											w.append(converted);
+											written.incrementAndGet();
+										} catch (AvroConversionException e) {
+											failures.add(e);
```

Comment:
This doesn't have to go in the initial version, but since we have the original schema and the original datum, we can send just the failed records on the "incompatible" relationship and not the entire file. I think that would be valuable, but I can add that when I add it to CSV and JSON.